[XAM/NUI] - Improve NUI Handling

The-Little-Wolf opened the pull request:
- Improve NUI handling
- Stub XamUnk2B001
- Stub XamUnk2B003
- Add XamUnk2B004 used in Emerald Dashboard
- Replaced unk parameters

has207 commented on src/xenia/kernel/nui.h:
```diff
@@ -0,0 +1,64 @@
+  uint64_t session_id = 0x0;
+  uint32_t nui_callback = 0x0;
+  uint32_t nui_hud_flags = 0x0;
+  uint64_t nui_ver_id[2] = {};
```
you have these member variables public but also define getters/setters, they should either be private or set/get them directly, having two ways to interact with them is kind of weird

has207 commented on src/xenia/kernel/nui.cc:
```diff
@@ -0,0 +1,35 @@
+uint32_t NUIController::GetHudFlags() { return nui_hud_flags; }
+void NUIController::SetHudFlags(uint32_t flags) { nui_hud_flags = flags; }
+uint64_t NUIController::GetNUIVerID(uint32_t index) {
+  return nui_ver_id[index];
```
missing bounds check

has207 commented on src/xenia/kernel/nui.h:
```diff
@@ -0,0 +1,64 @@
+#include <array>
+#include <cstdint>
+#include <functional>
+#include <map>
```
array, functional and map aren't used afaict, should remove the includes

has207 commented on src/xenia/kernel/nui.cc:
```diff
@@ -0,0 +1,35 @@
+ */
+
+#include "xenia/kernel/nui.h"
+#include "xenia/kernel/util/shim_utils.h"
```
is anything used from this include?

has207 commented on src/xenia/kernel/xam/xam_nui.cc:
```diff
@@ -123,7 +129,7 @@ dword_result_t XamNuiCameraTiltGetStatus_entry(lpvoid_t unk) {
      - Ghidra decompile fails
   */
-  return X_E_FAIL;
+  return X_ERROR_SUCCESS;
```
the unk buffer is never updated despite success return, it should at the minimum be zeroed I think, but also maybe good to add an XELOGD here so that it's not just faking this silently

has207 commented on src/xenia/kernel/xam/apps/xam_app.cc:
```diff
@@ -162,11 +163,24 @@ X_HRESULT XamApp::DispatchMessageSync(uint32_t message, uint32_t buffer_ptr,
         xe::be<uint64_t> unk1;
         xe::be<uint64_t> unk2;
-        xe::be<uint64_t> unk3;
+        xe::be<uint32_t> flags;
```
this changes the struct size, did you verify the new size is correct? It might cause some subtle alignment issues

The-Little-Wolf replied:
This is how its structured in ghidra when viewing xam 12625

has207 commented on src/xenia/kernel/nui.h:
```diff
@@ -0,0 +1,64 @@
+}  // namespace kernel
+}  // namespace xe
+
+#endif
```
#endif  // XENIA_KERNEL_NUI_H_

has207 commented on src/xenia/kernel/xam/xam_nui.cc:
```diff
@@ -106,14 +107,19 @@ dword_result_t XamNuiGetDepthCalibration_entry(lpdword_t unk1) {
 // Skeleton
-qword_result_t XamNuiSkeletonGetBestSkeletonIndex_entry(int_t unk) {
-  return 0xffffffffffffffff;
+int_result_t XamNuiSkeletonGetBestSkeletonIndex_entry(int_t unk) {
```
changing the return type from 64 to 32bit could produce wrong register state on the guest side, was this intentional / verified correct?

The-Little-Wolf committed:
[XAM/NUI] - Improve NUI

- Improve NUI handling
- Stub XamUnk2B003 used in
- Add XamUnk2B004 used in Emerald Dashboard
- Replaced unk parameters

The-Little-Wolf committed:
[XAM/NUI] - Improve NUI

- Improve NUI handling
- Stub XamUnk2B001
- Stub XamUnk2B003
- Add XamUnk2B004 used in Emerald Dashboard
- Replaced unk parameters
- Add X_NUI_DEPTH_CALIBRATION and X_NUI_PLAYER_ENGAGEMENT_UPDATE struct
